### DISCUSSION TITLE
Default argument heuristic for template parameters

### DISCUSSION THREAD

**Discussion by HighCommander4** [clangd/clangd#1056] [Ideas]
Clangd uses heuristics (implemented in the class `HeuristicResolver`) to try to resolve dependent names in templates to their declarations.

Currently, the main heuristic that's employed is assuming that a dependent template specialization will resolve to the primary template rather than to an explicit or partial specialization, which then allows us to heuristically resolve e.g. `A<T>::foo` (where `T` is dependent) by looking up `foo` in the scope of the primary template definition of `A`.

Another heuristic that's occurred to me recently is that if `T` is a template parameter, and it has a default argument, heuristic resolution could try to replce it with its default argument. So, e.g. if you're trying to resolve `Allocator::pointer_type` inside `vector<T, Allocator>`, currently we can't resolve that, but with this heuristic, if `Allocator` has the default arugment `std::allocator<T>`, we'd heuristically resolve it by replacing `Allocator::pointer_type` with `std::allocator<T>::pointer_type` (and then apply the existing heuristic to look up `pointer_type` in the primary template definition of `std::allocator`).

Thoughts? Does this heuristic seem useful / desirable?

**Comment by tom-anders** [comment]
This sounds really useful e.g. for dependency injection via templates:

```
class Implementation { 
   void foo();
 }

template <class Imp = Implementation>
class C {
   C(const Imp& imp) {
      imp.f^ // <- Would be cool if autocompletion would complete Imp::foo() here!
   }
}
```

**Reply by ops2abt** [reply]
This would be very useful! Is there any update if this will be implemented or if there is a workaround to still get suggestions in that use-case?

**Reply by HighCommander4** [reply]
I think this heuristic would be pretty simple to implement.

If anyone's interested in giving it a try, I'm happy to provide pointers to the relevant code and other guidance as needed.

**Reply by ops2abt** [reply]
Unfortunately, I don't really have the time and know-how to dive deep into the clangd code base, so you could see it as a feature request for know.

Maybe you could already write here a small abstract where the relevant code would need to be implemented and maybe some bullet points how you would address it / what the most important things are to consider there?

Then I or someone else might do it in the future :)

**Reply by HighCommander4** [reply]
As always, it's good to start with a test case. Here is a simple one:

```c++
struct Default {
  void foo();
};
template <typename T = Default>
void bar(T t) {
  t.foo();
}
```

With the heuristic in place, go-to-definition on the `foo` token in the function body of `bar` should take you to the declaration of `Default::foo()`.

This test case can be added to the test [`LocateSymbol.All`](https://searchfox.org/llvm/rev/cc53ec82ea6df6e7602510fa1bf5b8a991b3bc39/clang-tools-extra/clangd/unittests/XRefsTests.cpp#447); you can look at the existing test cases there to see how to formulate it.

Note that we're testing go-to-definition because that's the main feature that uses `HeuristicResolver` at this time. Code completion does not yet use `HeuristicResolver` (though we have https://github.com/clangd/clangd/issues/810 on file for that, and that's something I'm hoping to pick up this year) and so won't currently be affected.

As for the fix, I imagine it going around [here](https://searchfox.org/llvm/rev/cc53ec82ea6df6e7602510fa1bf5b8a991b3bc39/clang-tools-extra/clangd/HeuristicResolver.cpp#56), and involving the following steps:
 * check if `T` is a `TemplateTypeParmType`
 * if so, get its associated `TemplateTypeParmDecl` 
    * (note the caveat mentioned in [this comment](https://github.com/llvm/llvm-project/pull/71279#issuecomment-1818429734): the decl needs to be queried **before** unwrapping the type to its canonical type)
 * check if the `TemplateTypeParmDecl` has a default argument
 * if so, replace `T` with the type of the default argument

We can imagine making the heuristic a bit smarter (e.g. handling the case where the default argument ends up itself being a template parameter of another type, and applying the heuristic recursively in that case), but that can be left for a follow-up. What I outlined above should be sufficient to get the test case I wrote working.

**Comment by ops2abt** [comment]
Yes, it would be very useful. Actually, we just switched from IntelliSense to Clangd and would not switch back due to many great features that clangd provides. But this feature is really missing for us in clangd. We also use inl files to make the implementation separate from the template definition and then you have to jump to another file to find the place where you can jump to the Implementation...

What we currently do as workaround is, that we change the template in the inl file to \<class Implementation> and then you get the right suggestions. Then when we are finished with programming we switch it back to \<class Imp = Implementation>.

So this feature would really benefit us!

**Comment by HighCommander4** [comment]
I implemented this at https://github.com/llvm/llvm-project/pull/131074

**Reply by ops2abt** [reply]
Thank you! Works perfectly and that really helps us now :)

**Reply by Gnomiro** [reply]
@HighCommander4  Nice work! What else is necessary to make use of it? Is it already merged in the proper places to be part of a clangd release? I'm not very familiar with how the integration/release strategy works here.

**Reply by HighCommander4** [reply]
> @HighCommander4 Nice work! What else is necessary to make use of it? Is it already merged in the proper places to be part of a clangd release? I'm not very familiar with how the integration/release strategy works here.

Clangd is part of the LLVM project which puts out a release twice a year. You can read about the release schedule [here](https://llvm.org/docs/HowToReleaseLLVM.html#annual-release-schedule).

This change will appear in the LLVM 21 release which is expected in early September.

If you'd like to use it sooner, we also publish weekly snapshot releases at https://github.com/clangd/clangd/releases (and for Linux users, https://apt.llvm.org/ publishes nightly releases).
